drivers/bme680/bme680_spi_saul.c

```diff
@@ -0,0 +1,190 @@
+        _temp_valid[dev_index] = false;
+
+        data->val[0] = _temp[dev_index];
+        data->unit = UNIT_TEMP_C;
+        data->scale = -2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_press(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_spi_t *)dev);
+    if (dev_index == BME680_SPI_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_press_valid[dev_index] || _read(dev_index) == 0) {
+        /* mark local variable as invalid */
+        _press_valid[dev_index] = false;
+
+        data->val[0] = _press[dev_index];
+        data->unit = UNIT_PA;
+        data->scale = 2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_hum(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_spi_t *)dev);
+    if (dev_index == BME680_SPI_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_hum_valid[dev_index] || _read(dev_index) == 0) {
+        /* mark local variable as invalid */
+        _hum_valid[dev_index] = false;
+
+        data->val[0] = _hum[dev_index];
+        data->unit = UNIT_PERCENT;
+        data->scale = -2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_gas(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_spi_t *)dev);
+    if (dev_index == BME680_SPI_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_gas_valid[dev_index] || _read(dev_index) == 0) {
+        /* mark local variable as invalid */
+        _gas_valid[dev_index] = false;
+
+        if (_gas[dev_index] > INT16_MAX) {
+            data->val[0] = _gas[dev_index] / 1000;
+            data->scale = 3;
+        }
+        else {
+            data->val[0] = _gas[dev_index];
+            data->scale = 0;
+        }
+        data->unit = UNIT_OHM;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+const saul_driver_t bme680_spi_saul_driver_temperature = {
+    .read = read_temp,
+    .write = saul_notsup,
+    .type = SAUL_SENSE_TEMP
+};
+
+const saul_driver_t bme680_spi_saul_driver_pressure = {
+    .read = read_press,
+    .write = saul_notsup,
+    .type = SAUL_SENSE_PRESS
+};
+
+const saul_driver_t bme680_spi_saul_driver_humidity = {
+    .read = read_hum,
+    .write = saul_notsup,
+    .type = SAUL_SENSE_HUM
+};
+
+const saul_driver_t bme680_spi_saul_driver_gas = {
+    .read = read_gas,
+    .write = saul_notsup,
+    .type = SAUL_SENSE_GAS
+};
```

Comment:
I'm not sure if I understand that part correctly. It expects that both `spi_devs` and `i2c_devs` are stored in one array, right? For now there is one array for each one of them; and I can't see how to combine them.

Reply:
The suggested helper function actually relies on SPI and I2C being stored in separate arrays. If both modules `bme680_spi` and `bme680_i2c` are used, it will check if the given pointer points to memory within the SPI device array and returns `true` if this is the case.

The idea is that `bme680_spi` and `bme680_i2c` remain fully distinct, but that the SAUL implementation is part of `bme680_common` and works for both. Note that the return value of the helper function can be determined at compile time, if only on of the flavors is used. That way, the compiler will drop the dead branch. E.g. the `_read()` function could look similar to this

```C
static int _read(const void *unknown_dev)
{
    int res;
    bme680_data_t;
    if (_is_spi_attached(unknown_dev)) {
        const bme680_spi_t *dev = unknown_dev;
        res = bme680_spi_read(dev, &data);
    }
    else {
        const bme680_i2c_t *dev = unknown_dev;
        res = bme680_i2c_read(dev, &data);
    }
    ...
}
```